File: 2022/days/04/main.py
def getSpan(elfRange):
    return int(elfRange[2]) - int(elfRange[0]) + 1


def orderBySize(left, right):
    if getSpan(left) < getSpan(right):
        return left, right
    return right, left


def getOverlap(left, right):
    # mispredicted the problemn, so we calculate the overlap but in main loop just check we have any overlap
    # 0 1 2 3 4 5 6 7
    #     x-----x
    #   x---x
    #         x---x
    #       x-x
    # x-x
    #             x-x
    smaller, larger = orderBySize(left, right)

    # close out cases where our range is fully excluded
    if smaller[2] < larger[0] or smaller[0] > larger[2]:
        return 0

    # close out cases where we're fully enclosed
    if contained(smaller, larger):
        return getSpan(smaller)

    # our smaller is further to the left
    if smaller[0] <= larger[0] and smaller[2] <= larger[2]:
        return smaller[2] - larger[0] + 1

    # our smaller is further to the right
    if smaller[0] >= larger[0] and smaller[2] >= larger[2]:
        return larger[2] - smaller[0]  + 1

    return -1

def contained(left, right):
    smaller,larger = orderBySize(left,right)
    return (smaller[0] >= larger[0] and smaller[2] <= larger[2])

def process(elfRange):
    l,r = elfRange.split('-')
    return [int(l),'-', int(r)]
    

if __name__ == '__main__':
    file = 'input.txt'

    partOneAcc,partTwoAcc = 0,0
    with open(file, 'r') as fp:
        for line in fp:
            left, right = line.split(',')
            right = right.strip('\n')

            left,right = process(left),process(right)

            # handle part 1
            if contained(left,right):
                partOneAcc += 1

            # handle part 2
            if getOverlap(left,right) > 0:
                partTwoAcc += 1

    print(partOneAcc,partTwoAcc)

File: 2022/days/04/main.ts
import { readFileSync } from 'fs'


function processElfRange(elfRange: string) : number[]
{
    const stringRange: String[] = elfRange.split('-')
    return [Number(stringRange[0]),Number(stringRange[1])]
}

function elfRangesOrdered(left: number[],right: number[]): boolean {
    return elfRangeLength(left) <= elfRangeLength(right)
}

function elfRangeLength(elfRange:number[]): number{
    if(elfRange.length != 2){
        return -1
    }

    return (elfRange[1] - elfRange[0] + 1)
}

function getElfRangesOverlap(leftElfRange:number[],rightElfRange:number[]): number{
    let smaller = leftElfRange
    let larger = rightElfRange
    
    if (!elfRangesOrdered(smaller,larger)){
        [smaller,larger] = [rightElfRange,leftElfRange]
    }
    
    // cases:
    // smaller is outside our larger range
    // smaller is entirely inside our larger range
    // smaller intersects our larger range
    if (smaller[1] < larger[0] || smaller[0] > larger[1]){
        return 0
    }

    if(smaller[0] >= larger[0] && smaller[1] <= larger[1]){
        return elfRangeLength(smaller)
    }

    if(smaller[0]<larger[0]){
        return elfRangeLength([larger[0],smaller[1]])
    }

    return elfRangeLength([smaller[0],larger[1]])
}

const file = readFileSync('input.txt', 'utf-8')

let accPartOne: number = 0
let accPartTwo: number = 0
file.split('\n').forEach(pair => {
    let p = pair.split(',')
    p[1] = p[1].replace("\r","")

    let smaller :number[] = processElfRange(p[0])
    let larger :number[] = processElfRange(p[1])
    if(!elfRangesOrdered(smaller,larger)){
        [smaller,larger] = [larger,smaller]
    }

    let overlap:number = getElfRangesOverlap(smaller,larger)

    if(overlap == elfRangeLength(smaller)){
        accPartOne += 1
    }

    if (overlap != 0){
        accPartTwo +=1
    }
})

console.log(`${accPartOne},${accPartTwo}`)